### PR TITLE
Band member main

### PR DESCRIPTION
**Pull request by KevinHuang78759**
This pull request implements feature Foo. 

- Transferred code from BandMemberClass 2 branch to be mergable with main

### CODE REVIEW

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:295]: can you document this so we know at what rate this is changing and also explain what -1 means? I think it's worth making this some constant and less hard coded because the rate/how much it decreases will likely change.

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameMode.java:90]: not sure why you changed this

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameMode.java]: revert these changes because they're not true anymore.

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:38]: What happened with separating separate game updates (like the frames) from the GameplayController vs what makes the GameplayController? I think this was separated already.

This isn't the only case but if you don't see a need for this anymore then we can talk about this.

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java]: Please rename this to bandMembers because seeing `bms` is honestly hard to read. 

Also as a discussion point, this will get moved to a level class, correct?

**Review by mluo24 — COMMENTED** [review]
In terms of actual architecture, I will leave it up to Marya to comment, but there are a few things to note:

1. I left some general comments on the files/certain lines and I encourage you to review them and let me know if it makes sense or not.

2. You have a lot of public variables (variables without a labeled visibility). Can you change them to private unless you have a good reason to leave them public?

3. You don't have any competency bars draw on the screen, yet there is a lot of code that handles. Is this intended? At the very least I would recommend drawing the competency bar value.

4. This doesn't take into account beat/frame/song position based "beat" detection. I'm assuming Marya is going to handle this. 

5. There's a small bug I found when testing. I'm not too sure about the reproducibility steps, but I think I may have reset the level while switching by accident and this is what happened:
image
There are tiny notes that are falling on the lane, maybe because it's not registering we aren't in that lane anymore. 

Otherwise, everything else looks good to me and your code is significantly easier to follow than before. Good job!

I'll approve once I receive word from you and Marya approves.

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameMode.java:90]: so we don't have to use this.lpl = lpl;

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java]: Will do

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:38]: I was going to change that after pulling from main again

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameMode.java]: will do

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:295]: will do

**Comment by KevinHuang78759** [comment]
> In terms of actual architecture, I will leave it up to Marya to comment, but there are a few things to note:
> 
> 1. I left some general comments on the files/certain lines and I encourage you to review them and let me know if it makes sense or not.
> 2. You have a lot of public variables (variables without a labeled visibility). Can you change them to private unless you have a good reason to leave them public?
> 3. You don't have any competency bars draw on the screen, yet there is a lot of code that handles. Is this intended? At the very least I would recommend drawing the competency bar value.
> 4. This doesn't take into account beat/frame/song position based "beat" detection. I'm assuming Marya is going to handle this.
> 5. There's a small bug I found when testing. I'm not too sure about the reproducibility steps, but I think I may have reset the level while switching by accident and this is what happened:
> 
> image
> 
> There are tiny notes that are falling on the lane, maybe because it's not registering we aren't in that lane anymore.
> Otherwise, everything else looks good to me and your code is significantly easier to follow than before. Good job!
> 
> I'll approve once I receive word from you and Marya approves.

for 2. I added getters and setters to the Note and Bandmembers class already. Anything else that doesn't use a getter and setter is appropriate

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (mluo24) [reply core/src/edu/cornell/gdiac/temporary/GameMode.java:90]: if you looked at the diff it was `lpl = linesPerLane` before

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:295]: This method was already documented, it changes the bandmember's competency by the parameters in the field without exceeding any bounds

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (markim21) [reply core/src/edu/cornell/gdiac/temporary/GameMode.java:90]: 🫠

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/GameMode.java]: is this TODO for the future, or no longer in planning?

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/GameMode.java:228]: great that this is encapsulated in BandMembers class now

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:57]: Isn't health/competency
1. particular for each band member, and 
2. defined by the level? this should not be a `final int` or defined in `GameplayController`

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java]: not really significant, but variable names should have a consistent format. `randomNotes` instead of `randomnotes`

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:137]: naming

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:193]: as mentioned in my code before: shouldn't this be initialized in the bandMembers class?

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:5]: note not shell (again not really important ig)

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:20]: note descent should not be defined at the note level. it should be defined by the level or music level. also, shouldn't be final

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java]: `NoteType` but yeah my code doesn't really matter

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:109]: movement shouldn't be velocity based

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java]: note texture should be initialized here imo

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (markim21) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java]: typo I'm sure, `heightConfine`

**Review by markim21 — APPROVED** [review]
only had small comments, but overall lgtm

**Review by mluo24 — APPROVED** [review]
approving to streamline process so that you can merge stuff in once you address marya's changes

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:137]: To be fixed on the reimplementation of hp bars

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:193]: this is initializing bandMembers. We need to do this outside because the coordinates of each bandmember depend on the other bandmembers

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java:20]: Intended to be changed after this PR

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java]: Intended to be changed after this PR

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java:109]: Intended to be changed after this PR

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java]: Noted, will change

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java]: indeed

File: core/src/edu/cornell/gdiac/temporary/GameMode.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameMode.java]: What are you referring to here?
The data structure has essentially already been created

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:57]: Anything related to health or competency is to be changed after this PR

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java]: woops mb. Though, this variable isn't actually used anymore

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java:5]: fixed comments

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java
Comment (KevinHuang78759) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java]: I just looked at your NoteType, its the same as NType, but I will change the name

**Pull request merged**
